File: python/complex-numbers/complex_numbers.py
import math

class ComplexNumber(object):
    def __init__(self, real, imaginary):
        self.real = real
        self.imaginary = imaginary

    def __eq__(self, other):
        if self.real == other.real and self.imaginary == other.imaginary:
            return True
        else:
            return False

    def __add__(self, other):
        return self.__class__(self.real + other.real, self.imaginary + other.imaginary)

    def __mul__(self, other):
        return self.__class__(self.real * other.real - self.imaginary * other.imaginary,
                              self.real * other.imaginary + self.imaginary * other.real)

    def __sub__(self, other):
        return self.__class__(self.real - other.real, self.imaginary - other.imaginary)

    def __truediv__(self, other):
        real = (self.real * other.real + self.imaginary * other.imaginary) * 1.00 / (other.real ** 2 + other.imaginary ** 2)
        imaginary = (self.imaginary * other.real - self.real * other.imaginary) * 1.00 / (other.real ** 2 + other.imaginary ** 2)
        return self.__class__(real, imaginary)

    def __abs__(self):
        return math.sqrt(self.real ** 2 + self.imaginary ** 2)

    def conjugate(self):
        return self.__class__(self.real, -1 * self.imaginary)

    def exp(self):
        if self.imaginary == math.pi:
            real = math.e ** self.real * -1
            imaginary = math.e ** self.real * 0
            return self.__class__(real, imaginary)
        elif self.imaginary == 0:
            real = math.e ** self.real * 1
            imaginary = math.e ** self.real * 0
            return self.__class__(real, imaginary)

File: ruby/complex-numbers/complex_numbers.rb
class ComplexNumber
  include Comparable
  attr_reader :real, :imaginary

  def initialize(real, imaginary)
    @real = real
    @imaginary = imaginary
  end

  def *(other)
    self.class.new(@real * other.real - @imaginary * other.imaginary, @real * other.imaginary + @imaginary * other.real)
  end

  def /(other)
    real = (@real * other.real + @imaginary * other.imaginary) * 1.00 / (other.real ** 2 + other.imaginary ** 2)
    imaginary = (@imaginary * other.real - @real * other.imaginary) * 1.00 / (other.real ** 2 + other.imaginary ** 2)
    self.class.new(real, imaginary)
  end

  def +(other)
    self.class.new(@real + other.real, @imaginary + other.imaginary)
  end

  def -(other)
    self.class.new(@real - other.real, @imaginary - other.imaginary)
  end

  def <=>(other)
    (@real <=> other.real) && (@imaginary <=> other.imaginary)
  end

  def abs
    Math.sqrt(@real ** 2 + @imaginary ** 2)
  end

  def conjugate
    self.class.new(@real, -1 * @imaginary)
  end

  def exp
    if @imaginary == Math::PI
      real = Math::E ** @real * -1
      imaginary = Math::E ** @real * 0
      self.class.new(real, imaginary)
    elsif @imaginary == 0
      real = Math::E ** @real * 1
      imaginary = Math::E ** @real * 0
      self.class.new(real, imaginary)
    end
  end
end